File: chor-extraction/src/main/java/epp/MergingProjection.java
package epp;

import ast.sp.interfaces.Behaviour;
import ast.sp.interfaces.SPNode;
import ast.sp.nodes.*;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;


public class MergingProjection {
    public static SPNode merge(SPNode node1, SPNode node2) throws MergingException {
        try {
            Class<?> mp = MergingProjection.class;
            Method m = mp.getMethod("merge", node1.getClass(), node2.getClass());
            return (SPNode) m.invoke( null, node1, node2 );
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            throw new MergingException("Can't merge " + node1.getClass() + " with " + node2.getClass(), e);
        }
    }

    public static SPNode merge(TerminationSP left, TerminationSP right){
        return new TerminationSP();
    }

    public static SPNode merge(Sending left, Sending right) throws MergingException {
        assertCondition ((left.getProcess().equals(right.getProcess()))&&(left.getExpression().equals(right.getExpression())), "");
        Behaviour m = (Behaviour) merge(left.getContinuation(), right.getContinuation());
        return new Sending(m, left.getProcess(), left.getExpression());
    }

    public static SPNode merge(Receiving left, Receiving right) throws MergingException {
        assertCondition ((left.getProcess().equals(right.getProcess())), "");

        Behaviour m = (Behaviour) merge(left.getContinuation(), right.getContinuation());
        return new Receiving(m, left.getProcess());

    }

    public static SPNode merge(SelectionSP left, SelectionSP right) throws MergingException {
        assertCondition ((left.getProcess().equals(right.getProcess())),"");
        assertCondition ((left.getLabel().equals(right.getLabel())),"");

        Behaviour continuation = (Behaviour) merge(left.getContinuation(), right.getContinuation());
        return new SelectionSP(continuation, left.getProcess(), left.getLabel());
    }

    public static SPNode merge(Offering left, Offering right) throws MergingException {
        assertCondition((left.getProcess().equals(right.getProcess())), "" );

        HashMap<String, Behaviour> leftmap = left.getLabels();
        HashMap<String, Behaviour> rightmap = right.getLabels();
        HashMap<String, Behaviour> labels = new HashMap();

        for (String leftkey: leftmap.keySet()){
            if (rightmap.containsKey(leftkey)){
                Behaviour b = (Behaviour) merge(left.getLabels().get(leftkey), right.getLabels().get(leftkey));
                labels.put(leftkey, b);
                rightmap.remove(leftkey);
            } else {
                labels.put(leftkey, left.getLabels().get(leftkey));
            }

        }
        for (String rightkey: rightmap.keySet()){
            labels.put(rightkey, right.getLabels().get(rightkey));
        }

        return new Offering(left.getProcess(), labels);
    }

    public static SPNode merge(ConditionSP left, ConditionSP right) throws MergingException {
        Behaviour leftCondition = (Behaviour) merge(left.getThenBehaviour(), left.getThenBehaviour());
        Behaviour rightCondition = (Behaviour) merge(right.getElseBehaviour(), right.getElseBehaviour());
        assertCondition( left.getProcess().equals( right.getProcess() ), "" );
        assertCondition( left.getExpression().equals( right.getExpression() ), "" );
        return new ConditionSP(left.getProcess(), left.getExpression(), leftCondition, rightCondition );
    }

    public static SPNode merge(ProcedureInvocationSP left, ProcedureInvocationSP right){
        if (left.getProcedure().equals(right.getProcedure())) {
            return new ProcedureInvocationSP(left.getProcedure());
        } else return null;
    }

    public SPNode merge(ProcedureDefinitionSP left, ProcedureDefinitionSP right) throws MergingException {
        assertCondition( left.getProcedure().equals( right.getProcedure() ), "" );

        Behaviour mergeBehaviour = (Behaviour) merge(left.getBehaviour(), right.getBehaviour());
        return new ProcedureDefinitionSP(left.getProcedure(), mergeBehaviour);
    }

    private static void assertCondition( boolean condition, String message )
        throws MergingException {
        if ( !condition ) {
            throw new MergingException( message );
        }
    }
}

File: chor-extraction/src/main/java/epp/MergingProjection.kt
package epp

import ast.sp.interfaces.Behaviour
import ast.sp.interfaces.SPNode
import ast.sp.nodes.*
import java.lang.reflect.InvocationTargetException
import java.util.*


class MergingProjection {

    @Throws(MergingException::class)
    fun merge(left: ProcedureDefinitionSP, right: ProcedureDefinitionSP): SPNode {
        val mergeBehaviour = merge(left.behaviour, right.behaviour) as Behaviour
        return ProcedureDefinitionSP(mergeBehaviour)
    }

    companion object {
        @Throws(MergingException::class)
        fun merge(node1: SPNode, node2: SPNode): SPNode {
            try {
                val mp = MergingProjection::class.java
                val m = mp.getMethod("merge", node1.javaClass, node2.javaClass)
                return m.invoke(null, node1, node2) as SPNode
            } catch (e: NoSuchMethodException) {
                throw MergingException("Can't merge " + node1.javaClass + " with " + node2.javaClass, e)
            } catch (e: InvocationTargetException) {
                throw MergingException("Can't merge " + node1.javaClass + " with " + node2.javaClass, e)
            } catch (e: IllegalAccessException) {
                throw MergingException("Can't merge " + node1.javaClass + " with " + node2.javaClass, e)
            }

        }

        fun merge(left: TerminationSP, right: TerminationSP): SPNode {
            return TerminationSP()
        }

        @Throws(MergingException::class)
        fun merge(left: Sending, right: Sending): SPNode {
            assertCondition(left.process == right.process && left.expression == right.expression)
            val m = merge(left.continuation, right.continuation) as Behaviour
            return Sending(m, left.process, left.expression)
        }

        @Throws(MergingException::class)
        fun merge(left: Receiving, right: Receiving): SPNode {
            assertCondition(left.process == right.process)

            val m = merge(left.continuation, right.continuation) as Behaviour
            return Receiving(m, left.process)

        }

        @Throws(MergingException::class)
        fun merge(left: SelectionSP, right: SelectionSP): SPNode {
            assertCondition(left.process == right.process)
            assertCondition(left.expression == right.expression)

            val continuation = merge(left.continuation, right.continuation) as Behaviour
            return SelectionSP(continuation, left.process, left.expression)
        }

        @Throws(MergingException::class)
        fun merge(left: Offering, right: Offering): SPNode {
            assertCondition(left.process == right.process)

            val leftmap = left.labels
            val rightmap = right.labels
            val labels = HashMap<String, Behaviour>()

            for (leftkey in leftmap.keys) {
                if (rightmap.containsKey(leftkey)) {
                    val b = merge(left.labels[leftkey] as Behaviour, right.labels[leftkey] as Behaviour) as Behaviour
                    labels.put(leftkey, b)
                    rightmap.remove(leftkey)
                } else {
                    labels.put(leftkey, left.labels[leftkey] as Behaviour)
                }

            }
            for (rightkey in rightmap.keys) {
                labels.put(rightkey, right.labels[rightkey] as Behaviour)
            }

            return Offering(left.process, labels)
        }

        @Throws(MergingException::class)
        fun merge(left: ConditionSP, right: ConditionSP): SPNode {
            val leftCondition = merge(left.thenBehaviour, left.thenBehaviour) as Behaviour
            val rightCondition = merge(right.elseBehaviour, right.elseBehaviour) as Behaviour
            assertCondition(left.process == right.process)
            assertCondition(left.expression == right.expression)
            return ConditionSP(left.process, left.expression, leftCondition, rightCondition)
        }

        fun merge(left: ProcedureInvocationSP, right: ProcedureInvocationSP): SPNode? {
            if (left.procedure == right.procedure) {
                return ProcedureInvocationSP(left.procedure)
            } else
                return null
        }

        @Throws(MergingException::class)
        private fun assertCondition(condition: Boolean) {
            if (!condition) {
                throw MergingException("")
            }
        }
    }
}
